### PR TITLE
Write auth test

### LINKED ISSUE #16: Implement register user functionality

### LINKED ISSUE #19: Implement login and role-based navigation

### PR DESCRIPTION
**Pull request by nabekabebe**
fix #19 #16

### CODE REVIEW

**Review by par0s — APPROVED** [review]

File: lib/features/auth/bloc/signup/signup_form_bloc.dart [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+
+      if (state.status.isValidated) {
+        yield state.copyWith(status: FormzStatus.submissionInProgress);
+
+        final AuthSignUpParam param = AuthSignUpParam(
+            username: state.username.value,
+            email: state.email.value,
+            password: state.password.value);
+
+        final failureOrSuccess = await _authRepository.getRemoteProvider
```
Comment (surafel1999) [inline lib/features/auth/bloc/signup/signup_form_bloc.dart:92]: add createRemoteUser method for AuthRepository and call that function.

File: lib/features/auth/repository/repository.dart [hunk 2]
```diff
@@ -0,0 +1,20 @@
+import 'package:rental/features/auth/data_provider/local_provider.dart';
+import 'package:rental/features/auth/data_provider/remote_provider.dart';
+
+class AuthRepository {
```
Comment (surafel1999) [inline lib/features/auth/repository/repository.dart:4]: Wrap remote method calls in a method here

File: test/features/auth/bloc/signin/signin_bloc_test.dart [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+              status: FormzStatus.invalid,
+            )
+          ];
+        });
+
+    blocTest(
+        "emits [SignInFormState (valid), SignInFormState (In Progress), SignInFormState (success)] when form is submitted with valid inputs",
+        build: () {
+      final String token = "thisIsAFakeJWTTokenLongEnough";
+      when(mocksAuthRepository.getRemoteProvider)
```
Comment (surafel1999) [inline test/features/auth/bloc/signin/signin_bloc_test.dart:105]: Remove this stub

File: test/features/auth/bloc/signin/signin_bloc_test.dart [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+          .thenAnswer((_) => mockAuthRemoteDataProvider);
+      when(
+        mockAuthRemoteDataProvider.attemptLogin(
```
Comment (surafel1999) [inline test/features/auth/bloc/signin/signin_bloc_test.dart:108]: Change method call to authRepository.attemptRemoteLogin

File: test/features/auth/bloc/signup/signup_bloc_test.dart [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+              password: Password.pure("1fygf"),
+              status: FormzStatus.invalid,
+            )
+          ];
+        });
+
+    blocTest(
+        "emits [SignUpFormState (valid), SignUpFormState (In Progress), SignUpFormState (success)] when form is submitted with valid inputs",
+        build: () {
+      when(mocksAuthRepository.getRemoteProvider)
```
Comment (surafel1999) [inline test/features/auth/bloc/signup/signup_bloc_test.dart:132]: Remove this stub

File: test/features/auth/bloc/signup/signup_bloc_test.dart [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+          .thenAnswer((_) => mockAuthRemoteDataProvider);
+      when(mockAuthRemoteDataProvider.createUser(
```
Comment (surafel1999) [inline test/features/auth/bloc/signup/signup_bloc_test.dart:134]: Change to authRepository.createRemoteUser

**Review by seraph14 — CHANGES_REQUESTED** [review]
Address reviews in each file

**Review by natnaelabay — APPROVED** [review]

**Review by seraph14 — APPROVED** [review]

File: lib/features/auth/bloc/signup/signup_form_bloc.dart (hunk 1 above)
Comment (natnaelabay) [reply lib/features/auth/bloc/signup/signup_form_bloc.dart]: Incorrect

File: test/features/auth/bloc/signin/signin_bloc_test.dart (hunk 3 above)
Comment (natnaelabay) [reply test/features/auth/bloc/signin/signin_bloc_test.dart]: fix everything

File: test/features/auth/bloc/signup/signup_bloc_test.dart (hunk 6 above)
Comment (natnaelabay) [reply test/features/auth/bloc/signup/signup_bloc_test.dart]: @seraph14 Great review

**Pull request merged**